src/qtviewer/ParseIfcFile.cpp

```diff
@@ -21,22 +31,145 @@ void ParseIfcFile::Parse(const std::string& filePath)
 
     IfcGeom::Iterator* it = new IfcGeom::Iterator(settings, &file);
     if (!it->initialize()) {
-        outputMsg("Error: Iterator failed to initialize! Aborting.");
+        MessageLogger::log("Error: Iterator failed to initialize! Aborting.");
         delete it;
-        return;
+        return false;
     }
 
     do {
         //const IfcGeom::BRepElement* bRepElem = it->get_native();
         const IfcGeom::TriangulationElement* triElem = static_cast<const IfcGeom::TriangulationElement*>(it->get());
-        outputMsg(triElem->type() + ": " + triElem->name());
+        MessageLogger::log(triElem->type() + ": " + triElem->name());
 
         const boost::shared_ptr<IfcGeom::Representation::Triangulation>& triElemGeom = triElem->geometry_pointer();
 
+        const std::vector<int>& elemFaces = triElemGeom->faces();
+        const std::vector<double>& elemVertices = triElemGeom->verts();
+        const std::vector<double>& elemNormals = triElemGeom->normals();
+
+        std::vector<Triangle> triangles = this->createTriangles(elemFaces, elemVertices, elemNormals);
```

Comment:
This seems a bit inefficent to first compose everything into triangles and later on in the for loop on line 56 decompose everything again into arrays of vertices and normals. It also likely creates copies of the vertices, where previously a vertex could be referenced by multiple triangles.

Reply:
I agree, cycling through the triangles again could lead to an overhead in case of very complex geometries.
Implemented a revised flow.